src/wallet/coinselection.cpp

```diff
@@ -335,23 +338,29 @@ util::Result<SelectionResult> KnapsackSolver(std::vector<OutputGroup>& groups, c
         for (const auto& group : applicable_groups) {
             result.AddInput(group);
         }
-        return result;
+        if (result.GetWeight() <= max_input_weight) return result;
+
+        // Try something else
+        result.Clear();
     }
 
     if (nTotalLower < nTargetValue) {
-        if (!lowest_larger) return util::Error();
+        if (!lowest_larger) return util::Error{Untranslated(strprintf("Only %d coins to cover target value of %d", nTotalLower, nTargetValue))};
```

Comment:
hmmm, this change to give a useful error message seems to break tests; I start getting failures in rpc_psbt.py

Can I get some assistance on how this stuff is supposed to bubble up @murchandamus

Reply:
Is the breakage not an actual issue: https://cirrus-ci.com/task/6124189116006400?logs=ci#L3274
```bash
test_framework.authproxy.JSONRPCException: Only 4999872180 coins to cover target value of 5100000840 (-4)
```

Reply:
I can turn the error on and off locally by adding/removing the string from the error.

Reply:
> The empty util::Error() is used to retrieve a "no solution found for the provided inputs set". If you specify an error message, then the wallet coin selection procedure treats it as a real error and bubble it up when no other selection algorithm succeeded.

Ok. So then this is just a (confusing) misuse of the `Util::` API, where the wallet code is shoehorning non-error handling into an error handler.

Reply:
> > The empty util::Error() is used to retrieve a "no solution found for the provided inputs set". If you specify an error message, then the wallet coin selection procedure treats it as a real error and bubble it up when no other selection algorithm succeeded.
> 
> Ok. So then this is just a (confusing) misuse of the `Util::` API, where the wallet code is shoehorning non-error handling into an error handler.

Ideally, we should retrieve more info within the error return path and not just the string message. I was waiting until we expand the result class functionality to accept a generic error return type. E.g. having `util::Result<M, E>` would let us do something like `util::Result<SelectionResult, SelectionError>`, where `SelectionError` would contain the string msg and a severity level int value. Which would help us differentiate the different outcomes and return the highest severity error (if any). I actually left a comment for this in the [code](https://github.com/bitcoin/bitcoin/blob/e69796c79c0aa202087a13ba62d9fbcc1c8754d4/src/wallet/spend.cpp#L867).

Reply:
> I haven't gone deeper over this PR yet but the message you are adding here doesn't seems like an error. It is just describing the "no solution found for the provided input set" transient situation.

Are you saying it's not confusing to give insufficient funds message on failure to stay within the weight limits given? I guess that's fine for now?

Reply:
> > I haven't gone deeper over this PR yet but the message you are adding here doesn't seems like an error. It is just describing the "no solution found for the provided input set" transient situation.
> 
> Are you saying it's not confusing to give insufficient funds message on failure to stay within the weight limits given? I guess that's fine for now?

Not really. I was only referring to this line in question. Which, in a first glance, isn't retrieving an exceeded weight limit error. Still, give me a day to review the PR in-detail and will join the efforts here.

Reply:
still waiting on some feedback on this point before moving forward with the PR